tests/frontend/test_circuitspecs_chip0.py

```diff
@@ -0,0 +1,300 @@
+            assert cmd1.op.__class__ == cmd2.op.__class__
+            assert [p.x for p in cmd1.op.p] == [p.x for p in cmd2.op.p]
+            assert [i.ind for i in cmd1.reg] == [i.ind for i in cmd2.reg]
+
+    def test_not_all_modes_measured(self):
+        """Test exceptions raised if not all modes are measured"""
+        prog = sf.Program(4)
+        U = random_interferometer(2)
+
+        with prog.context as q:
+            ops.S2gate(0.5) | (q[0], q[2])
+            ops.S2gate(0.5) | (q[1], q[3])
+            ops.Interferometer(U) | (q[0], q[1])
+            ops.Interferometer(U) | (q[2], q[3])
+            ops.MeasureFock() | (q[0], q[1])
+
+        with pytest.raises(CircuitError, match="All modes must be measured"):
+            res = prog.compile("chip0")
+
+    def test_no_s2gates(self):
+        """Test exceptions raised if no S2gates are present"""
+        prog = sf.Program(4)
+        U = random_interferometer(2)
+
+        with prog.context as q:
+            ops.Interferometer(U) | (q[0], q[1])
+            ops.Interferometer(U) | (q[2], q[3])
+            ops.MeasureFock() | q
+
+        with pytest.raises(CircuitError, match="must start with two S2gates"):
+            res = prog.compile("chip0")
+
+    def test_incorrect_s2gates(self):
+        """Test exceptions raised if S2gates are on incorrect modes"""
+        prog = sf.Program(4)
+        U = random_interferometer(2)
+
+        with prog.context as q:
+            ops.S2gate(0.5) | (q[0], q[2])
+            ops.Interferometer(U) | (q[0], q[1])
+            ops.Interferometer(U) | (q[2], q[3])
+            ops.MeasureFock() | q
+
+        with pytest.raises(CircuitError, match="S2gates placed on the incorrect modes"):
+            res = prog.compile("chip0")
+
+    def test_no_unitary(self):
+        """Test compilation works with no unitary provided"""
+        prog = sf.Program(4)
+
+        with prog.context as q:
+            ops.S2gate(0.5) | (q[0], q[2])
+            ops.S2gate(0.5) | (q[1], q[3])
+            ops.MeasureFock() | q
+
+        res = prog.compile("chip0")
+
+        expected = sf.Program(4)
+
+        with expected.context as q:
+            ops.S2gate(0.5, 0) | (q[0], q[2])
+            ops.S2gate(0.5, 0) | (q[1], q[3])
+
+            # corresponds to an identity on modes [0, 1]
+            ops.Rgate(0) | q[0]
+            ops.BSgate(np.pi/4, np.pi/2) | (q[0], q[1])
+            ops.Rgate(np.pi) | q[0]
+            ops.BSgate(np.pi/4, np.pi/2) | (q[0], q[1])
+            ops.Rgate(np.pi) | q[0]
+            ops.Rgate(0) | q[1]
+
+            # corresponds to an identity on modes [2, 3]
+            ops.Rgate(0) | q[2]
+            ops.BSgate(np.pi/4, np.pi/2) | (q[2], q[3])
+            ops.Rgate(np.pi) | q[2]
+            ops.BSgate(np.pi/4, np.pi/2) | (q[2], q[3])
+            ops.Rgate(np.pi) | q[2]
+            ops.Rgate(0) | q[3]
+
+            ops.MeasureFock() | (q[0], q[3], q[1], q[2])

```

Comment:
Note: `GBSSpec.compile()` seems to permute the measured mode order. I don't _think_ this is an issue anywhere, but may be unintuitive to the user, and worth fixing in another PR.

Reply:
I seem to recall TFBackend gives an exception if modes are out of order in `measure_fock`

Reply:
Ah, so worthwhile pinning down and fixing then. I'm not really sure where the permutation happens, but my best guess is `GBSSpec.compile`

Reply:
seems likely it's happening within one of the calls to networkx

Reply:
The permutation probably happens during `lexicographical_topological_sort()`, since all measurement permutations are equivalent. The easiest fix is to add a `sorted()` call to `GBSSpec.compile()` using the subsystem index as the sorting key.